browser/ui/views/brave_rewards/tip_panel_bubble_host.cc

```diff
@@ -0,0 +1,144 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/ui/views/brave_rewards/tip_panel_bubble_host.h"
+
+#include <memory>
+#include <utility>
+
+#include "brave/browser/brave_rewards/rewards_util.h"
+#include "brave/browser/ui/webui/brave_rewards/tip_panel_ui.h"
+#include "brave/components/constants/webui_url_constants.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/ui/views/bubble/webui_bubble_manager.h"
+#include "chrome/browser/ui/views/frame/browser_view.h"
+#include "chrome/browser/ui/views/location_bar/location_bar_view.h"
+#include "components/grit/brave_components_strings.h"
+#include "ui/gfx/geometry/rounded_corners_f.h"
+#include "ui/views/controls/webview/webview.h"
+#include "url/gurl.h"
+
+namespace brave_rewards {
+
+namespace {
+
+// The tip dialog is anchored to the center of the location bar, and must be
+// drawn with rounded bottom borders. We implement these customizations by
+// subclassing `WebUIBubbleDialogView`. This implementation is then used by
+// `TipPanelBubbleManager` when creating the dialog view.
+class TipPanelDialogView : public WebUIBubbleDialogView {
+ public:
+  TipPanelDialogView(views::View* anchor_view,
+                     BubbleContentsWrapper* contents_wrapper,
+                     const absl::optional<gfx::Rect>& anchor_rect)
+      : WebUIBubbleDialogView(anchor_view, contents_wrapper, anchor_rect) {
+    SetArrow(views::BubbleBorder::TOP_CENTER);
+  }
+
+  // views::Widget:
+  std::unique_ptr<views::NonClientFrameView> CreateNonClientFrameView(
+      views::Widget* widget) override {
+    auto frame = WebUIBubbleDialogView::CreateNonClientFrameView(widget);
+    auto* bubble_frame = static_cast<views::BubbleFrameView*>(frame.get());
+    auto* bubble_border = bubble_frame->bubble_border();
+    DCHECK(bubble_border);
+    bubble_border->SetRoundedCorners(0, 0, 16, 16);
+    return frame;
+  }
+
+  // views::View:
+  void AddedToWidget() override {
+    WebUIBubbleDialogView::AddedToWidget();
+    web_view()->holder()->SetCornerRadii(gfx::RoundedCornersF(0, 0, 16, 16));
+  }
+};
+
+// In order to use `TipPanelDialogView` as the dialog view implementation, we
+// must override the `CreateWebUIBubbleDialog` of `WebUIBubbleManager`.
+class TipPanelBubbleManager : public WebUIBubbleManager {
+ public:
+  TipPanelBubbleManager(views::View* anchor_view, Profile* profile)
+      : anchor_view_(anchor_view), profile_(profile) {}
+
+  ~TipPanelBubbleManager() override = default;
```

Comment:
Delete copy/assignment as well?

Reply:
They are deleted in the subclass, which prevents copying. Do we need to explicitly delete in the subclass as a style thing?

Reply:
Nope, I think that's fine then actually.